README.md

```diff
@@ -1 +1,263 @@
+Ipld::DagJson::serialize(
+    Ipld::Assoc([
+        ("role", IPLD::String("user")),
+        (
+            "links",
+            Ipld::Array([
+                Ipld::Cid("bafkreidb2q3ktgtlm5yio7buj3sypyghjtfh5ernsteqmakf4p2c5bwmyi"),
+                Ipld::Cid("bafkreic75ydg5vkw324oqkcmqltfvc3kivyngqkibjoysdwiilakh4z5fe"),
+                Ipld::Cid("bafkreiffdiz6raf46zrr3b2usufgz5fo44aggmocz4zappr6khhhljcdpy"),
+            ]),
+        )
+    ])
+);
+```
+
+The signature is then checked against the above fields, which passes since there's only a `role: "user"` entry. The application then goes onto continue to use the original JSON with the `role: "admin"` entry.
+
+# 2 Safety
+
+Data that has already been parsed to an in-memory IPLD representation can be canonically encoded to a trivially: it has already been through a [parser / validator](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).
+
+Data purporting to conform to an IPLD encoding (such as [DAG-JSON](https://ipld.io/specs/codecs/dag-json/spec/)) MUST be validated prior to signature verification. This can be as simple as round-trip decoding/encoding the JSON and checking that the hash matches. A validation error MUST be signalled if it does not match. 
+
+> [Implementers] may provide an opt-in for systems where round-trip determinism is a desireable [sic] feature and backward compatibility with old, non-strict data is unnecessary.
+>
+> — [DAG-JSON Spec](https://ipld.io/specs/codecs/dag-json/spec/)
+
+As it is critical for signatures guard against various attacks, the assumptions around canonical encoding MUST be enforced.
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
+
+## 2.2 Raw (Noncanonicalized) Data
+
+Canonicalization is not required if data is encoded as raw bytes (multicodec `0x55`). The exact bytes are already present, and MUST not be changed.
+
+# 3 Varsig Format
+
+After being decoded from Varints, a varsig includes the following segments:
+
+| Segment Name              | Type          | Description                                     | Required |
+|---------------------------|---------------|-------------------------------------------------|----------|
+| Varsig Multiformat Prefix | `0x34`        | The multicodec varsig prefix                    | Yes      |
+| Content Multicodec Prefix | `Multiformat` | The IPLD encoding uses to canonicalize the data | Yes      |
+| Public Key Prefix         | `Multicodec`  | The multicodec prefix for the public key type   | Yes      |
+| Multihash                 | `Multihash`   | The multicodec prefix for the multihash         | Yes      |
+| Raw Signature             | `Bytes`       | The raw signature                               | Yes      |
```

Comment:
Hm, is there a way in IPLD to more directly bundle the `"Multihash"` and `"Raw Signature"` fields together, so that they're not two separate fields, but instead one field that carries along with it the expected length and format of the signature?

I don't see a concrete attack, but something makes me squeamish about the Varsig parser needing to know the lengths of each signature for decoding purposes.

Reply:
@QuinnWilton this is how multiformats tend to get formatted. They use varints (LEB128) to let you know how many bytes are in each segment even if they're tightly packed.

Does that help assuage your concerns?

Reply:
> Ah or @QuinnWilton I may have misunderstood the concern. Do you mean like the expected length of an RSA 1024- vs 2048-bit signature?

Yup, I mean the length of the raw signature. I hadn't seen [this](https://github.com/ChainAgnostic/varsig/pull/6#discussion_r1040280417) comment thread before, but if I'm understanding correctly, it's about the same question, so I'll move discussion there :)